### PR TITLE
Feature/cosmos

### PR DESCRIPTION
**Pull request by ProofOfKeags**

### CODE REVIEW

**Comment by keepkeyjon** [comment]
This PR points at a different commit than https://github.com/keepkey/keepkey-firmware/pull/188/files

File: messages-stellar.proto [hunk 1]
```diff
@@ -0,0 +1,223 @@
+syntax = "proto2";
```
Comment (keepkeyjon) [inline messages-stellar.proto]: Stellar stuff should go in its own PR.

File: messages-cosmos.proto [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    required uint64 account_number = 2;
+    required string chain_id = 3;
+    required uint32 fee_amount = 4;
+    required uint32 gas = 5;
+    required string from_address = 6;
+    required string to_address = 7;
+    required uint64 amount = 8;
+    required string memo = 9;
+    required uint64 sequence = 10;
+}
```
Comment (keepkeyjon) [inline messages-cosmos.proto:53]: In the other commit, you've got:

https://github.com/keepkey/device-protocol/compare/88e29410753b26d27882ed66460e8bd61d6980c6...2e297ce5c9550ee72015f9f607359bae15266113#diff-59a453ca48d4fcab5089a5bcfea9247fR37

... which I like. But let's take that a step further:

```
message CosmosSignTx {
    repeated uint32 address_n = 1;
    optional uint64 account_number = 2;
    optional string chain_id = 3; // Is this really a `string`, or should it be `optional bytes chain_id = 3;` ?
    optional CosmosFee fee = 4;
    optional string memo = 5; // We may want to stick this in a CosmosMsgAck, so we can support arbitrary length memos.
    optional uint64 sequence = 6;
    optional uint32 msg_count = 7;
}

message CosmosMsgRequest {
}

message CosmosMsgAck {
    optional CosmosMsgSend send = 1;
    // This leaves room in the protocol for adding more cosmos messages later
}

message CosmosMsgSend {
    optional string from_address = 1; // Do we actually need this, or can we derive it via address_n + user's keys?
    optional string to_address = 2;
    optional uint64 amount = 3;
    // This implicitly assumes we're sending ATOMs. Maybe we should make room for cosmos tokens?
}
```

Sidenote: we should use `optional` everywhere, and do checking in the firmware, since `optional` gives us significantly more upgrade flexibility than `required`.

File: messages-cosmos.proto [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+/**
+ * Response: signature for transaction
+ * @end
+ */
+message CosmosSignedTx {
+    optional bytes public_key = 1;  // public key for the private key used to sign data
+    optional bytes signature = 2;   // signature suitable for sending to the Stellar network
```
Comment (keepkeyjon) [inline messages-cosmos.proto]: stellar copy-pasta

File: messages-cosmos.proto (hunk 2 above)
Comment (keepkeyjon) [reply messages-cosmos.proto:53]: > // We may want to stick this in a CosmosMsgAck, so we can support arbitrary length memos.

We talked about this internally, and I think it's fine to leave the memo as a field on `CosmosSignTx`.

Comment (ProofOfKeags) [reply messages-cosmos.proto:53]: am I correct in saying that `CosmosMsgAck` needs to be added to `CosmosSignTx` or am I missing something about how we are going to get the message stream?

Comment (keepkeyjon) [reply messages-cosmos.proto:53]: `*Ack` messages are the host responding to the corresponding `*Request` message that comes from the device. That means, for a two-output transaction, you'd see this on the wire:

```
host -> device: CosmosSignTx { ... msg_count=2 }
device -> host: CosmosMsgRequest {}
host -> device: CosmosMsgAck { send: { /* details for output 1 */ } }
device -> host: CosmosMsgRequest {}
host -> device: CosmosMsgAck { send: { /* details for output 2 */ } }
device -> host: CosmosSignedTx {...}
```

And each time the device gets a `CosmosMsgAck`, it will confirm that output (or whatever it ends up being when we add more cosmos msg types) with the user, and then append the relevant data to the incremental hash. Once it has all the pieces, it returns the signature.

Comment (keepkeyjon) [reply messages-cosmos.proto:53]: This is a bit different than how you've currently architected the protocol, which will mean some changes to the firmware and python bits.... which is why I wanted to chat about the protocol design *before* you got it all implemented.

Comment (keepkeyjon) [reply messages-cosmos.proto:53]: As a reference, you can look at the EOS implementation, which does something similar to allow an arbitrary number of actions in a single tx.

Comment (ProofOfKeags) [reply messages-cosmos.proto:53]: Yeah unfortunately the only time we had before the deadline (today) to do this was Sat/Sun, so we couldn't really talk about it prior. I'll take a look at EOS.

File: messages-cosmos.proto [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+}
+
+/**
+ * Request: ask device to sign Cosmos transaction
+ * @start
+ * @next CosmosSignedTx
+ */
+message CosmosSignTx {
+    repeated uint32 address_n = 1;          // BIP-32 path. For compatibility with other wallets, must be m/44'/118'/index'
+    optional uint64 account_number = 2;
```
Comment (keepkeyjon) [inline messages-cosmos.proto]:
```suggestion
    optional uint64 account_number = 2 [jstype = JS_STRING];
```

File: messages-cosmos.proto [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    optional string chain_id = 3;
+    optional uint32 fee_amount = 4;
+    optional uint32 gas = 5;
+    optional string memo = 6;
+    optional uint64 sequence = 7;
```
Comment (keepkeyjon) [inline messages-cosmos.proto]:
```suggestion
    optional uint64 sequence = 7 [jstype = JS_STRING];
```

File: messages-cosmos.proto [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+message CosmosMsgRequest {}
+
+message CosmosMsgAck {
+    optional CosmosMsgSend send = 1;
+}
+
+message CosmosMsgSend {
+    optional string from_address = 6;
+    optional string to_address = 7;
+    optional uint64 amount = 8;
```
Comment (keepkeyjon) [inline messages-cosmos.proto]:
```suggestion
    optional uint64 amount = 8 [jstype = JS_STRING];
```

**Review by keepkeyjon — APPROVED** [review]
Some small nits that we will need when we go to implement this in hdwallet, but LGTM otherwise.

File: messages-cosmos.proto (hunk 5 above)
Comment (ProofOfKeags) [reply messages-cosmos.proto]: could you explain what this does?

Comment (keepkeyjon) [reply messages-cosmos.proto]: When we generate the javascript/typescript bindings, this specifies the type used to represent this field. Since `number` in js is really a float64, it can't precisely represent integers that are larger than will fit in the mantissa... but we can represent it with a string.

see: https://github.com/protocolbuffers/protobuf/issues/3666

Comment (keepkeyjon) [reply messages-cosmos.proto]: This shouldn't affect the C/Python bindings.

File: messages-cosmos.proto [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    optional string address = 1;    // Address Bech32 encoded with 'cosmos' prefix
+}
+
+/**
+ * Request: ask device to sign Cosmos transaction
+ * @start
+ * @next CosmosSignedTx
+ */
+message CosmosSignTx {
+    repeated uint32 address_n = 1;          // BIP-32 path. For compatibility with other wallets, must be m/44'/118'/index'
```
Comment (keepkeyjon) [inline messages-cosmos.proto:33]:
```suggestion
    repeated uint32 address_n = 1;          // BIP-32 path. For compatibility with other wallets, must be m/44'/118'/index'/0/0
```
